src/ir/module-splitting.cpp

```diff
@@ -0,0 +1,405 @@
+/*
+ * Copyright 2020 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ir/module-splitting.h"
+#include "ir/manipulation.h"
+#include "ir/module-utils.h"
+#include "ir/names.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace ModuleSplitting {
+
+namespace {
+
+std::unique_ptr<Module> initializeSecondary(const Module& primary) {
+  // Create the secondary module and copy trivial properties.
+  auto secondary = std::make_unique<Module>();
+  secondary->features = primary.features;
+  secondary->hasFeaturesSection = primary.hasFeaturesSection;
+  return secondary;
+}
+
+void shareImportableItems(Module& primary,
+                          Module& secondary,
+                          Name importNamespace) {
+  // Map internal names to (one of) their corresponding export names.
+  std::unordered_map<Name, Name> exports;
+  for (auto& ex : primary.exports) {
+    if (ex->kind != ExternalKind::Function) {
```

Comment:
why not functions? no obvious reason so the reader is forced to look lower down, unless I'm missing something obvious.

Reply:
Functions are exported only as necessary because we expect there to be many of them compared to memories, tables, globals, and events. I'll add a comment.